bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -760,15 +783,25 @@ boolean refreshImageForZoom () {
 		}
 	} else {
 		if (deviceZoomLevel != currentDeviceZoom) {
-			ImageData data = getImageDataAtCurrentZoom();
-			destroy ();
-			ImageData resizedData = DPIUtil.autoScaleImageData(device, data, deviceZoomLevel, currentDeviceZoom);
-			init(resizedData);
-			init();
-			refreshed = true;
-			currentDeviceZoom = deviceZoomLevel;
+			if (this.dataAtBaseZoomFactor == null && memGC == null) {
+				// Cache data at base zoom before refresh.
+				this.dataAtBaseZoomFactor = getImageData(currentDeviceZoom);
+				this.dataBaseZoomFactor = this.currentDeviceZoom;
+			}
+			if (this.dataAtBaseZoomFactor != null) {
+				ImageData resizedData = getImageData(deviceZoomLevel);
+				destroy ();
+				init(resizedData);
+				init();
+				refreshed = true;
+				currentDeviceZoom = deviceZoomLevel;
+			}
 		}
 	}
+	if (refreshed) {
+		// Reset cached width and height to -1, if the image is changed
+		width = height = -1;
+	}
```

Comment:
Isn't it already a bug in the current implementatation that these values are not reset upon rescaling the image?

Reply:
But that is the method where the image is rescaled. Where would you expect the reset of width and height?

Reply:
I rephrase my question: why is resetting `width` and `height` necessary when the image is refreshed now (i.e., after this PR) and was not necessary before? My understanding is that this method still initialized the image with data for a different zoom factor, like before. So the `width` and `height` will usually change and are recalculated upon calling `getBoundsInPixels()`.

Reply:
Ah okay, sorry I missed your point before. I need to have a short look at it. I don't remember the scenario out of my head